flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/source/split/IcebergSourceSplit.java

```diff
@@ -0,0 +1,128 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink.source.split;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.connector.source.SourceSplit;
+import org.apache.iceberg.CombinedScanTask;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.flink.source.Position;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Objects;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+
+@Internal
+public class IcebergSourceSplit implements SourceSplit, Serializable {
+  private final CombinedScanTask task;
+  /**
+   * Position field is mutable
+   */
+  @Nullable
+  private final Position position;
+
+  /**
+   * The splits are frequently serialized into checkpoints.
+   * Caching the byte representation makes repeated serialization cheap.
+   */
+  @Nullable
+  private transient byte[] serializedFormCache;
+
+  public IcebergSourceSplit(CombinedScanTask task, Position position) {
+    this.task = task;
+    this.position = position;
+  }
+
+  public static IcebergSourceSplit fromCombinedScanTask(CombinedScanTask combinedScanTask) {
+    return fromCombinedScanTask(combinedScanTask, 0, 0L);
+  }
+
+  public static IcebergSourceSplit fromCombinedScanTask(
+      CombinedScanTask combinedScanTask, int fileOffset, long recordOffset) {
+    return new IcebergSourceSplit(combinedScanTask, new Position(fileOffset, recordOffset));
+  }
+
+  public CombinedScanTask task() {
+    return task;
+  }
+
+  public Position position() {
+    return position;
+  }
+
+  byte[] serializedFormCache() {
+    return serializedFormCache;
+  }
+
+  void serializedFormCache(byte[] cachedBytes) {
+    this.serializedFormCache = cachedBytes;
+  }
+
+  @Override
+  public String splitId() {
+    return MoreObjects.toStringHelper(this)
+        .add("files", toString(task.files()))
+        .toString();
+  }
+
+  public void updatePosition(int newFileOffset, long newRecordOffset) {
+    position.update(newFileOffset, newRecordOffset);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    IcebergSourceSplit split = (IcebergSourceSplit) o;
+    return Objects.equal(splitId(), split.splitId()) &&
+        Objects.equal(position, split.position());
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(splitId());
```

Comment:
Using the value of `toString()` to calculate the hasCode is not a good practice because if some `FileScanTask` implementation did not implement the correct `toString()`, it will use the instance's hashCode as its `toString()` value.  That leads to the meaningless of comparing two `IcebergSourceSplit`.

Reply:
As we have the cached `serializedFormCache`, why not use the lazy approach to get the serialized bytes and calculate its byte array's hashCode ?

Reply:
I don't think that we have a guarantee that `serializedFormCache` will be set unless the split has in fact been serialized via `IcebergSourceSplitSerializer`.

Reply:
Then why not assign an unique integer number to the `IcebergSourceSplit` as `splitId` when planing the tasks in `FlinkSplitPlanner#planIcebergSourceSplits` ?  I think keeping the `toString` approach as the identifier did not answer the question I raised in the first comment.

Reply:
I still don't think it valid to make the hash code build on top of the `FileScanTask#toString` because this [comment](https://github.com/apache/iceberg/pull/3501/files#r745462545).